### PR TITLE
Replace lapack with linalg in MahalanobisDistance

### PR DESCRIPTION
**Pull request by vinx13**
This PR replaces lapack usage with linalg in `MahalanobisDistance`.
A unit test of `MahalanobisDistance` is added (reference computed with Matlab)

### CODE REVIEW

**Comment by karlnapf** [comment]
Great!

File: cmake/FindMetaExamples.cmake [hunk 1]
```diff
@@ -19,7 +19,6 @@ function(get_excluded_meta_examples)
     read_config_header(${shogun_INCLUDE_DIR}/shogun/lib/config.h)
     is_set(${SHOGUN_CONFIG} HAVE_NLOPT)
     is_set(${SHOGUN_CONFIG} USE_GPL_SHOGUN)
-    is_set(${SHOGUN_CONFIG} HAVE_LAPACK)
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: we want to keep that, there might be other examples

File: cmake/FindMetaExamples.cmake [hunk 2]
```diff
@@ -37,13 +36,6 @@ function(get_excluded_meta_examples)
 			)
 	ENDIF()
 
-	IF(NOT HAVE_LAPACK)
-		LIST(APPEND EXCLUDED_META_EXAMPLES
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: please keep the block, so we can add things there that pop up in the future, just don't append anything

File: src/shogun/regression/LinearRidgeRegression.cpp [hunk 3]
```diff
@@ -8,7 +8,6 @@
  */
 #include <shogun/lib/config.h>
 
-#ifdef HAVE_LAPACK
```
Comment (karlnapf) [inline src/shogun/regression/LinearRidgeRegression.cpp]: ah so this guard was just outdated?

I think we should use linalg rather than eigen3 directly inside though

Comment (vinx13) [reply src/shogun/regression/LinearRidgeRegression.cpp]: Yes, just outdated. I will look into that to find the possibility of using pure linalg.

**Comment by vinx13** [comment]
both `MahalanobisDistance` and `LinearRidgeRegression` are covered by integration tests.

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 3 above)
Comment (karlnapf) [reply src/shogun/regression/LinearRidgeRegression.cpp]: leave it as it is for now (you can already change the rest)
Then you could create this method in linalg, and then send another PR

**Comment by karlnapf** [comment]
I think the LAPACK guard was still necessary.
BTW if you remove such guards, make sure that Shogun compiles if the library that you guarded is not available. Really important.

The windows build doesnt have lapack installed, and here you go:
```
[00:29:53]   C:\projects\shogun\src\shogun\distance\MahalanobisDistance.cpp(56): error C2039: 'inverse': is not a member of 'shogun::SGMatrix<float64_t>' [C:\projects\shogun\build\src\shogun\libshogun.vcxproj]
[00:29:53]   C:\projects\shogun\src\shogun\distance\MahalanobisDistance.cpp(56): error C3861: 'inverse': identifier not found [C:\projects\shogun\build\src\shogun
```

**Comment by karlnapf** [comment]
It is this line:
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/distance/MahalanobisDistance.cpp#L58

I am sure that can be done with linalg.
But actually this is bad generally. One should never ever store matrix inverse. Instead, it should store a cholesky factorization of the covariance matrix. Then, rather than multiplying the inverse covariance matrix to a vector in the `compute` method, you should do a cholesky (triangular) solve. This is implemented in linalg

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 3 above)
Comment (karlnapf) [reply src/shogun/regression/LinearRidgeRegression.cpp]: Or, you could just add a loop for now.

**Comment by vinx13** [comment]
@karlnapf Since cov matrix is PSD, we should use [LDLT](https://eigen.tuxfamily.org/dox/classEigen_1_1MatrixBase.html#a0ecf058a0727a4cab8b42d79e95072e1) (linalg is using LLT) for Cholesky decomposition, right? Shall I add it to linalg?

**Comment by karlnapf** [comment]
The matrix is PSD indeed, so LDLT would be more stable.
https://eigen.tuxfamily.org/dox/group__TutorialLinearAlgebra.html

I think adding an argument to cholesky would be good, so that the factorization can operate in both modes, something like or `use_ldlt` which by default is false. Also make sure to update the docs appropriately so it is clear what is used.

Good catch!

**Comment by vinx13** [comment]
Seems simply adding a `use_ldlt` option will not work. In LDLT, we want both L and D. One solution is to overload `cholesky_factor` as `SGMatrix<T> cholesky_factor(const SGMatrix<T>& A, SGMatrix<T> &D)`. Since D is diagonal, alternatively we could use a SGVector. Besides, we may need to add a LDLT solver.

**Comment by karlnapf** [comment]
You are absolutely right. In fact I now remember a discussion we previously had about this some years back. I think a clean solution would be to expose a new method and a few solver method as well. Everything else is too messy.

**Comment by OXPHOS** [comment]
We did have some discussion years ago. The best way seems to be having separate methods for `LLT` and `LDLT`. And since now the `linalg` methods are split into several files, having more than one cholesky decomposition or solver would be less messy...

**Comment by vinx13** [comment]
I see. I'll try to implement that.

File: src/shogun/distance/MahalanobisDistance.h [hunk 4]
```diff
@@ -10,44 +10,45 @@
...
+	 * \f]
+	 *
+	 * i.e., instead of the mean as reference two vector \f$x_i\f$ and
+	 * \f$x_i'\f$
+	 * are compared.
+	 *
+	 * @see <a href="http://en.wikipedia.org/wiki/Mahalanobis_distance">
+	 * Wikipedia: Mahalanobis Distance</a>
+	 */
+	class CMahalanobisDistance : public CRealDistance
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.h]: BTW why this reformatting?
Pls don't put this into a PR that changes small bits as this one. Makes everything much harder to review

File: src/shogun/regression/LinearRidgeRegression.cpp [hunk 5]
```diff
@@ -72,28 +68,16 @@ bool CLinearRidgeRegression::train_machine(CFeatures* data)
...
-    Map<MatrixXd> eigen_kernel_matrix(kernel_matrix.matrix, num_feat,num_feat);
-    Map<MatrixXd> eigen_feats_matrix(feats_matrix.matrix, num_feat,num_vec);
-    Map<VectorXd> eigen_y(y.vector, num_feat);
-    Map<VectorXd> eigen_labels(((CRegressionLabels*)m_labels)->get_labels(),num_vec);
-    Map<VectorXd> eigen_tau(tau_vector.vector, num_feat);
-
-    eigen_kernel_matrix = eigen_feats_matrix*eigen_feats_matrix.transpose();
-
-    eigen_kernel_matrix.diagonal() += eigen_tau;
+	linalg::matrix_prod(feats_matrix, feats_matrix, kernel_matrix, false, true);
```
Comment (karlnapf) [inline src/shogun/regression/LinearRidgeRegression.cpp]: nice much cleaner

File: src/shogun/regression/LinearRidgeRegression.cpp [hunk 6]
```diff
@@ -72,28 +68,16 @@ bool CLinearRidgeRegression::train_machine(CFeatures* data)
...
 
-    LLT<MatrixXd> llt;
-    llt.compute(eigen_kernel_matrix);
-    if(llt.info() != Eigen::Success)
-    {
-    	SG_WARNING("Features covariance matrix was not positive definite\n");
-    	return false;
-    }
-    eigen_y = llt.solve(eigen_y);
+	auto decomposition = linalg::cholesky_factor(kernel_matrix);
```
Comment (karlnapf) [inline src/shogun/regression/LinearRidgeRegression.cpp]: why did you change the indentation? Is this the style cheker?

**Comment by karlnapf** [comment]
Could you pls split the PR into two: one for regression and one for the distance.
You still haven't checked whether shogun compiles with the guard remove and no lapack installed.
Pls don't remove guards without checking whether things still work: https://ci.appveyor.com/project/vigsterkr/shogun/build/1873#L6478

File: src/shogun/distance/MahalanobisDistance.h (hunk 4 above)
Comment (vinx13) [reply src/shogun/distance/MahalanobisDistance.h]: Seems the style checker changed other codes than mine. I'm using command like `git clang-format-3.8 --commit xxx` , is it right ?

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 6 above)
Comment (vinx13) [reply src/shogun/regression/LinearRidgeRegression.cpp]: Just noticed that the original code is indented with spaces.

**Comment by vinx13** [comment]
@karlnapf Created a new PR for `LinearRidgeRegression` (#4149 ). I will wait until #4130 be merged before I can continue with the refactor of `MahalanobisDistance`

**Comment by karlnapf** [comment]
yep, you can just force update this one now

**Comment by karlnapf** [comment]
But you will need to fix the "inverse" problem (see above) first.

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 7]
```diff
@@ -38,21 +36,25 @@ CMahalanobisDistance::~CMahalanobisDistance()
...
 {
 	CRealDistance::init(l, r);
-
+	SGMatrix<float64_t> cov;
 
 	if ( l == r)
 	{
 		mean = ((CDenseFeatures<float64_t>*) l)->get_mean();
-		icov  = ((CDenseFeatures<float64_t>*) l)->get_cov();
+		cov = ((CDenseFeatures<float64_t>*)l)->get_cov();
```
Comment (vigsterkr) [inline src/shogun/distance/MahalanobisDistance.cpp]: while you are changing these lines, could you please replace these explicit casts to `dynamic_cast` at least plz? :)
or even better do a `REQUIRED` assertation for the type

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 8]
```diff
@@ -38,21 +36,25 @@ CMahalanobisDistance::~CMahalanobisDistance()
...
 	}
 	else
 	{
 		mean = ((CDenseFeatures<float64_t>*)l)
 		           ->compute_mean((CDotFeatures*)lhs, (CDotFeatures*)rhs);
-		icov = CDotFeatures::compute_cov((CDotFeatures*) lhs, (CDotFeatures*) rhs);
+		cov = CDotFeatures::compute_cov((CDotFeatures*)lhs, (CDotFeatures*)rhs);
```
Comment (vigsterkr) [inline src/shogun/distance/MahalanobisDistance.cpp]: another casting

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 9]
```diff
@@ -83,12 +85,8 @@ float64_t CMahalanobisDistance::compute(int32_t idx_a, int32_t idx_b)
...
 		diff[i] = bvec.vector[i] - diff[i];
 
-	SGVector<float64_t> v = diff.clone();
-	cblas_dgemv(CblasColMajor, CblasNoTrans,
-		icov.num_rows, icov.num_cols, 1.0, icov.matrix,
-		diff.vlen, diff.vector, 1, 0.0, v.vector, 1);
-
-	float64_t result = cblas_ddot(v.vlen, v.vector, 1, diff.vector, 1);
+	auto v = linalg::ldlt_solver(chol_cov_L, chol_cov_d, chol_cov_p, diff);
+	float64_t result = linalg::dot(v, diff);
```
Comment (vigsterkr) [inline src/shogun/distance/MahalanobisDistance.cpp]: this could be auto or? :)

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 10]
```diff
@@ -39,20 +37,43 @@ bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
 {
 	CRealDistance::init(l, r);
 
+	REQUIRE(
+	    lhs->get_feature_class() == C_DENSE,
+	    "Left hand side (was %s) has to be CDenseFeatures instance.\n",
+	    lhs->get_name());
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: this segfaults if one of the features is not set (or does the above call check that?)

Comment (karlnapf) [reply src/shogun/distance/MahalanobisDistance.cpp]: Also: "Left hand side feature type (%s) has be be (%s)\n"
One should not use type names in the messages (think refactor), would be better to have a (static) method CDenseFeatures::get_name(). I'll send a patch and you can use that

Comment (vigsterkr) [reply src/shogun/distance/MahalanobisDistance.cpp]: https://github.com/vinx13/shogun/blob/39e7dd2f390d13e665678be23b0b0221f1212033/src/shogun/distance/Distance.cpp#L80

Comment (vigsterkr) [reply src/shogun/distance/MahalanobisDistance.cpp]: but the dense checks can be dropped as they are implicitly tested in `CRealDistance::init` and belove

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 11]
```diff
@@ -39,20 +37,43 @@ bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
...
+	REQUIRE(
+	    rhs->get_feature_class() == C_DENSE,
+	    "Right hand side (was %s) has to be CDenseFeatures instance.\n",
+	    rhs->get_name());
+
+	REQUIRE(
+	    lhs->get_feature_type() == F_DREAL,
+	    "Left hand side (was %s) has to be of double type.\n", lhs->get_name());
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: "Left hand side feature type (%d) has to be REAL"

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 12]
```diff
@@ -39,20 +37,43 @@ bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
 {
 	CRealDistance::init(l, r);
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: this call covers all the checks, so we don't need them again in this call (just convolutes the code)
@vigsterkr both REAL and DENSE is checked. That was the whole point of having the base class here iirc

Comment (karlnapf) [reply src/shogun/distance/MahalanobisDistance.cpp]: I would do a cast after the call and add a comment

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 13]
```diff
@@ -39,20 +37,43 @@ bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
...
+	REQUIRE(
+	    lhs->get_feature_class() == C_DENSE,
+	    "Left hand side (was %s) has to be CDenseFeatures instance.\n",
+	    lhs->get_name());
+	REQUIRE(
+	    rhs->get_feature_class() == C_DENSE,
+	    "Right hand side (was %s) has to be CDenseFeatures instance.\n",
+	    rhs->get_name());
+
+	REQUIRE(
```
Comment (vigsterkr) [inline src/shogun/distance/MahalanobisDistance.cpp]: and actually this is implicitly tested in `CRealDistance::init`
https://github.com/vinx13/shogun/blob/39e7dd2f390d13e665678be23b0b0221f1212033/src/shogun/distance/RealDistance.h#L34

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 14]
```diff
@@ -63,9 +84,10 @@ void CMahalanobisDistance::cleanup()
 
 float64_t CMahalanobisDistance::compute(int32_t idx_a, int32_t idx_b)
 {
+	auto feat_l = static_cast<CDenseFeatures<float64_t>*>(lhs);
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: this is reliable due to old code design principles by sören

File: src/shogun/distance/MahalanobisDistance.cpp (hunk 12 above)
Comment (vigsterkr) [inline src/shogun/distance/MahalanobisDistance.cpp]: instead of the REQUIRED sections it should be just check the return type of `CRealDistance::init(l, r)` as it checks for everything you need want....

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 15]
```diff
@@ -83,17 +105,13 @@ float64_t CMahalanobisDistance::compute(int32_t idx_a, int32_t idx_b)
 	for (int32_t i=0; i < diff.vlen; i++)
 		diff[i] = bvec.vector[i] - diff[i];
 
-	SGVector<float64_t> v = diff.clone();
-	cblas_dgemv(CblasColMajor, CblasNoTrans,
-		icov.num_rows, icov.num_cols, 1.0, icov.matrix,
-		diff.vlen, diff.vector, 1, 0.0, v.vector, 1);
-
-	float64_t result = cblas_ddot(v.vlen, v.vector, 1, diff.vector, 1);
+	auto v = linalg::ldlt_solver(chol_cov_L, chol_cov_d, chol_cov_p, diff);
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: yeah this is much much from a numerical perspective!

File: src/shogun/distance/MahalanobisDistance.cpp (hunk 13 above)
Comment (karlnapf) [reply src/shogun/distance/MahalanobisDistance.cpp]: haha I wrote the same thing above ;) sorry didnt see

File: src/shogun/distance/MahalanobisDistance.cpp (hunk 12 above)
Comment (vigsterkr) [reply src/shogun/distance/MahalanobisDistance.cpp]: i would actually change the design of this and have instead of init returning a boolean the actual typecasted values available... as i dont see really the point of an assertation failure that will just throw an exception and in case of success have a true :)
this function call should actually tell us that yes the types are cool and here you go this is what you wanted anyways

Comment (vigsterkr) [reply src/shogun/distance/MahalanobisDistance.cpp]: @vinx13 you can ignore most of this... just make sure that you do an ASSERT(CRealDistance::init(l, r)); and add a //FIXME: above with reference to the above comment

Comment (karlnapf) [reply src/shogun/distance/MahalanobisDistance.cpp]: @vigsterkr I mean this whole type system in the features is messed up. I would just rely on `dynamic_cast` everywhere. The bool is also pointless. And the enums are even worse (too many types, I would make those strings or even better drop them)

**Comment by vinx13** [comment]
@karlnapf could you plz take a look

File: src/shogun/distance/MahalanobisDistance.cpp [hunk 16]
```diff
@@ -37,22 +35,31 @@ CMahalanobisDistance::~CMahalanobisDistance()
 
 bool CMahalanobisDistance::init(CFeatures* l, CFeatures* r)
 {
-	CRealDistance::init(l, r);
+	// FIXME: See comments in
```
Comment (karlnapf) [inline src/shogun/distance/MahalanobisDistance.cpp]: fixme?
